File: backend/api/odds.py
from __future__ import annotations

import asyncio

from fastapi import APIRouter, HTTPException

from db import db_connection

router = APIRouter()


@router.get("/odds/{match_id}")
async def latest_odds(match_id: str):
    def load():
        with db_connection() as conn:
            row = conn.execute(
                """
                SELECT
                  match_id,
                  provider,
                  bookmaker,
                  home_odds,
                  draw_odds,
                  away_odds,
                  home_implied_prob,
                  draw_implied_prob,
                  away_implied_prob,
                  overround,
                  timestamp
                FROM odds_snapshots
                WHERE match_id = %s
                ORDER BY timestamp DESC, id DESC
                LIMIT 1
                """,
                (match_id,),
            ).fetchone()
        return row

    row = await asyncio.to_thread(load)
    if not row:
        raise HTTPException(status_code=404, detail="odds not found")

    return {
        "data": {
            "match_id": row[0],
            "provider": row[1],
            "bookmaker": row[2],
            "home_odds": float(row[3]),
            "draw_odds": float(row[4]),
            "away_odds": float(row[5]),
            "implied_probability": {
                "home": float(row[6]),
                "draw": float(row[7]),
                "away": float(row[8]),
            },
            "overround": float(row[9]),
            "timestamp": row[10].isoformat(),
        }
    }

File: services/api/src/routes/odds.ts
import { Router } from "express";

import { query } from "../db.js";
import { asyncHandler } from "../utils/asyncHandler.js";

export const oddsRouter = Router();

oddsRouter.get(
  "/:matchId",
  asyncHandler(async (req, res) => {
    const result = await query(
      `SELECT
         match_id,
         provider,
         bookmaker,
         home_odds,
         draw_odds,
         away_odds,
         home_implied_prob,
         draw_implied_prob,
         away_implied_prob,
         overround,
         timestamp
       FROM odds_snapshots
       WHERE match_id = $1
       ORDER BY timestamp DESC, id DESC
       LIMIT 1`,
      [req.params.matchId]
    );

    const row = result.rows[0];
    if (!row) {
      res.status(404).json({ error: { code: "odds_not_found", message: "odds not found" } });
      return;
    }

    res.json({
      data: {
        matchId: row.match_id,
        provider: row.provider,
        bookmaker: row.bookmaker,
        homeOdds: Number(row.home_odds),
        drawOdds: Number(row.draw_odds),
        awayOdds: Number(row.away_odds),
        marketProbabilities: {
          home: Number(row.home_implied_prob),
          draw: Number(row.draw_implied_prob),
          away: Number(row.away_implied_prob)
        },
        overround: Number(row.overround),
        timestamp: new Date(row.timestamp).toISOString()
      }
    });
  })
);
